File: Codility/18_MaxProductOfThree/solution.cpp
#include <iostream>
#include <vector>
#include <algorithm>

using namespace std;

int solution(vector<int>&A){
    sort(A.begin(), A.end());

    int n = A.size();
    int last_three = A[n-1] * A[n-2] * A[n-3];
    int first_two_and_last_one = A[0] * A[1] * A[n-1];

    return max(last_three, first_two_and_last_one);
}

int main(){
    vector<int> A = {-5, -10, 0, 1, 3, 5, 2};
    cout << "Max product is: " << solution(A) << endl;
    return 0;
}

File: Codility/18_MaxProductOfThree/solution.py
# nC3은 경우가 매우 많으니 정렬을 사용
def solution(A):
    sorted_list = sorted(A)
    
    last_three = sorted_list[-1] * sorted_list[-2] * sorted_list[-3]
    
    first_two_and_last_one = sorted_list[-1] * sorted_list[0] * sorted_list[1]
    
    return max(last_three, first_two_and_last_one)

# O(N * log(N))